.collapse('hide') on hidden uninit-ed collapsible no longer shows it

cvrebert opened the pull request:
Fixes #15315.
Closes #15807.
X-Ref: #14282, #14417

CC: @XhmikosR @hnrch02 @juthilo for review

cvrebert committed:
.collapse('hide') on hidden uninit-ed collapsible no longer shows it; fixes #15315

Thanks to @peterblazejewicz & @Nikita240
Adds unit tests based on #14417
X-Ref: #14282
Closes #15807

cvrebert commented:
Damn. The one old test this makes fail [dates back to the advent of the Collapse plugin's test suite](https://github.com/twbs/bootstrap/commit/1ef5fa7d6b4e50230c0c12919b0a06a9a2ac07f1#diff-4773e851610b64b333c3075ef5c248b0R22), so we might need to ask @fat about this.
I'm not sure why it's testing for adding `style="height: 0px;"`, considering that `.collapse` without `.in` sets `display: none` anyway...

hnrch02 commented on js/tests/unit/collapse.js:
```diff
@@ -177,6 +177,34 @@ $(function () {
+    setTimeout(done, 0)
+  })
+
+  QUnit.test('should hide a collapse when initialized with "hide" if not already hidden', function (assert) {
```
Maybe change the name to "should hide a collapse when initialized with "hide" *option* if not already hidden"

cvrebert replied:
I'm basically just following how the analogous "show" tests above these are phrased

hnrch02 replied:
Okay.

cvrebert replied:
Done.

hnrch02 commented on js/tests/unit/collapse.js:
```diff
@@ -177,6 +177,34 @@ $(function () {
+    assert.expect(1)
+    var done = assert.async()
+
+    $('<div class="collapse in"></div>')
```
$('<div class="collapse in"/>')

cvrebert replied:
That's not valid HTML and such shortcuts have caused some problems in the past (I think it with IE8, but still).

hnrch02 replied:
`grep "/>" js/tests/unit/*` and https://github.com/twbs/bootstrap/pull/13853#discussion_r14574124

hnrch02 replied:
We can change it for the whole suite if you feel strongly about it. Can't remember any problem with IE8 though.

hnrch02 commented on js/tests/unit/collapse.js:
```diff
@@ -177,6 +177,34 @@ $(function () {
+    $('<div class="collapse"></div>')
+      .appendTo('#qunit-fixture')
+      .on('show.bs.collapse', function () {
+        assert.ok(false, 'showing a previously-uninitialized hidden collapse when "hide" method called')
```
Again, not a native speak but shouldn't it be "showing a previously-uninitialized hidden collapse when "hide" method *is* called"

kkirsche replied:
I would agree with that @hnrch02

cvrebert replied:
Rephrased

hnrch02 commented:
I think it's save to remove/rework that old test.

cvrebert committed:
Collapse unit test: rm check that no-op hide adds style="height: 0px"

Appears to be unnecessary and is broken by #16011,
which is otherwise quite sane.

cvrebert commented:
Ok, took the plunge and removed that one old assertion.

cvrebert commented:
@XhmikosR LGTY?

XhmikosR commented:
Makes sense. :+1:
